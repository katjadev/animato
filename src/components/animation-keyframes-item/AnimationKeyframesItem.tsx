import { FC } from 'react'
import { REM_TO_PX_COEFFICIENT, TIMELINE_PADDING } from '@animato/constants'
import findClosestTimelineMark from '@animato/utils/findClosestTimelineMark'
import { useEditorState } from '@animato/components/editor/EditorContextProvider'
import { AnimationKeyframe } from '@animato/types'
import styles from './AnimationKeyframesItem.module.css'

interface AnimationKeyframesItemProps {
  keyframe: AnimationKeyframe;
  prevKeyframe?: AnimationKeyframe;
}

const AnimationKeyframesItem: FC<AnimationKeyframesItemProps> = ({ keyframe, prevKeyframe }) => {
  const { state, actions } = useEditorState()
  const { timelineMarks } = state

  const offset = TIMELINE_PADDING * REM_TO_PX_COEFFICIENT
  const position = findClosestTimelineMark(timelineMarks, keyframe.time).position + offset
  const prevPosition = prevKeyframe 
    ? findClosestTimelineMark(timelineMarks, prevKeyframe.time).position + offset
    : 0
  
  const handleClick = () => actions.setCurrentTime({ time: keyframe.time })

  return (
    <>
      <button 
        className={styles.keyframe}
        aria-label={`Keyframe: ${keyframe.time} milliseconds`}
        style={{ left: `${position}px` }}
        onClick={handleClick}
      />
      {prevKeyframe && (
        <div 
          className={styles.keyframeLine}
          style={{
            left: `${prevPosition}px`,
            width: `${position - prevPosition}px`,
          }}
        />
      )}
    </>
  )
}

export default AnimationKeyframesItem
