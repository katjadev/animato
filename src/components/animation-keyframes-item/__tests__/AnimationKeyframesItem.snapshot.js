import { render } from '@testing-library/react'
import AnimationKeyframesItem from '../AnimationKeyframesItem'

describe('AnimationKeyframesItem', () => {
  it('renders correctly without previous keyframe', () => {
    const keyframe = { time: 1000 }

    const { container } = render(<AnimationKeyframesItem keyframe={keyframe} />)
    
    expect(container).toMatchSnapshot()
  })

  it('renders correctly with previous keyframe', () => {
    const keyframe = { time: 1000 }
    const prevKeyframe = { time: 500 }
    
    const { container } = render(
      <AnimationKeyframesItem 
        keyframe={keyframe} 
        prevKeyframe={prevKeyframe} 
      />
    )
    expect(container).toMatchSnapshot()
  })
})
