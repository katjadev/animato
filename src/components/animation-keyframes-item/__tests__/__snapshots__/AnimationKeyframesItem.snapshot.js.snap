// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnimationKeyframesItem renders correctly with previous keyframe 1`] = `
<div>
  <button
    aria-label="Keyframe: 1000 milliseconds"
    class="keyframe"
    style="left: 176px;"
  />
  <div
    class="keyframeLine"
    style="left: 96px; width: 80px;"
  />
</div>
`;

exports[`AnimationKeyframesItem renders correctly without previous keyframe 1`] = `
<div>
  <button
    aria-label="Keyframe: 1000 milliseconds"
    class="keyframe"
    style="left: 176px;"
  />
</div>
`;
